crates/devolutions-agent-shared/src/windows/registry.rs

```diff
@@ -0,0 +1,86 @@
+use crate::windows::uuid::{reversed_hex_to_uuid, uuid_to_reversed_hex, UuidError};
+use crate::DateVersion;
+
+const REG_CURRENT_VERSION: &str = "SOFTWARE\\Microsoft\\Windows\\CurrentVersion";
+
+#[derive(Debug, thiserror::Error)]
+pub enum RegistryError {
+    #[error("failed to open registry key `{key}`")]
+    OpenKey { key: String, source: std::io::Error },
+    #[error("failed to enumerate registry key values from `{key}`")]
+    EnumKeyValues { key: String, source: std::io::Error },
+    #[error("failed to read registry value `{value}` from key `{key}`")]
+    ReadValue {
+        value: String,
+        key: String,
+        source: std::io::Error,
+    },
+    #[error(transparent)]
+    Uuid(#[from] UuidError),
+}
+
+/// Get the product code of an installed MSI using its upgrade code.
+pub fn get_product_code(update_code: &str) -> Result<Option<String>, RegistryError> {
+    let reversed_hex_uuid = uuid_to_reversed_hex(update_code)?;
+
+    let key_path = format!("{REG_CURRENT_VERSION}\\Installer\\UpgradeCodes\\{reversed_hex_uuid}");
+
+    let update_code_key = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE).open_subkey(&key_path);
+
+    // Product not installed if no key found.
+    let update_code_key = match update_code_key {
+        Ok(key) => key,
+        Err(_) => return Ok(None),
+    };
+
+    // Product code is the name of the only value in the registry key.
+    let (product_code, _) = match update_code_key.enum_values().next() {
+        Some(value) => value.map_err(|source| RegistryError::EnumKeyValues { key: key_path, source })?,
+        None => return Ok(None),
+    };
+
+    Ok(Some(reversed_hex_to_uuid(&product_code)?))
```

Comment:
question: What about working with the `Uuid` type instead of `String`s? Is there a reason for not doing that?

Reply:
suggestion: For the hex-reversed representation, you could define a wrapper type too.
See this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=641a51af984214092e0c365d07c0af3e

Reply:
Windows "reversed UUID" representation, as I call it, is pretty strange, and reverses UUID byte groups of different size per-hex character, not per-byte, e.g:
`C3D81328F118D5D4A9287B3CB1707655` => `{82318d3c-811f-4d5d-9a82-b7c31b076755}`
`C3D81328 => 82318d3c`
`F118` => `811f`

So we can't use from_str_radix with base16 here, and if we use UUID instead of string, then we still need to format `Uuid` to `String`,

I am not sure that switching to `uuid` will improve things here - right after we parsed UUID from string read from registry, we need to immediately format it to string back for conversion (or reverse in binary splitting hex data on 4-bit nibbles which is even more messy) 🤔

Reply:
Discussed in slack, switched to using Uuid for non-reversed form